README.md
# Guidance and framework for running HPC applications on Azure

This repository provides automation scripts for creating [Azure
Batch](https://azure.microsoft.com/services/batch/) pools that you can use to run common high-performance
computing (HPC) applications. This repo also serves as a catalog of HPC
applications that you can use for testing. More than a dozen common
HPC applications are currently supported, including several ANSYS solvers and
Star-CCM+, and you can add more as needed as described in this guide.

**In this guide:**

- [Overview](#overview)
    - [Azure Batch considerations](#azure-batch-considerations)
- [Prerequisites](#prerequisites)
    - [Set up NFS server and network](#set-up-nfs-server-and-virtual-network)
    - [Create site-to-site VPN](#create-site-to-site-vpn)
- [Create Azure resources and
  configuration](#create-azure-resources-and-configuration)
    - [Create a Key Vault to store secrets](#create-a-key-vault-to-store-secrets)
    - [Specify subscription and location](#specify-subscription-and-location)
    - [Get Packer](#get-packer)
    - [Store the images](#store-the-images)
    - [Specify the cluster type](#specify-the-cluster-type)
    - [Identify infrastructure addresses](#identify-infrastructure-addresses)
    - [Set up analytics](#set-up-analytics)
    - [Application storage](#application-storage)
    - [Store results](#store-results)
- [Run application](#run-application)
    - [Set up an HPC application](#set-up-an-hpc-application)
    - [Add an application to the catalog](#add-an-application-to-the-catalog)
    - [Add test cases](#add-test-cases)
    - [MPI environment variables](#mpi-environment-variables)
    - [Log data](#log-data)
- [Collect telemetry data](#collect-telemetry-data)

# Overview

The goal of this repo is to help you set up an environment in Azure
using either [Azure Batch](https://azure.microsoft.com/services/batch/) pools. These compute job scheduling services connect to an HPC application
and run it efficiently on Azure so you can perform tests.

As the following figure shows, this readme describes a three-step approach:

1.  Build an image. The automation scripts use [HashiCorp
    Packer](https://www.packer.io/), an open source tool for creating build
    images. Custom images are stored in separate infrastructure blobs.

2.  Create a Batch pool to automate the creation of the
    compute nodes (virtual machines) used to run your HPC application.

3.  Run the application. Pools of virtual machines are created with a custom
    image on demand. You can optionally set up Azure Log Analytics to collect
    telemetry during the testing.

![architecture](documentation/images/architecture.png)

## Azure Batch considerations

Batch offers clustering as a service that includes an integrated scheduler.
Batch creates and manages a pool of virtual machines , installs the applications
you want to run, and schedules jobs to run on the nodes. You only pay for the
underlying resources consumed, such as the virtual machines, storage, and
networking. 

# Fast start

Before using these scripts, do the following:

1.  It is highly recommended to use [Cloud Shell](https://azure.microsoft.com/en-us/features/cloud-shell/) which is the fastest way to start. 

> Notes:
>
>If you decide to use a centos VM then you will need to 
>- install jq (yum install -y jq)
>- install azure cli and do "az login"
>    - https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest
>- install packer (As of this time we recommend that you install 1.2.5)
>    - https://releases.hashicorp.com/packer/1.2.5/
>
>If running Bash on Windows, follow these
>    [instructions](https://www.michaelcrump.net/azure-cli-with-win10-bash/).

2.  Clone the repo:
```
    git clone git@github.com:az-cat/az-hpcapps.git
```

3. Fill up the configuration file by running the setup script
```
    ./setup.sh
```

4.  Install [Batch Explorer](https://github.com/Azure/BatchExplorer) for monitoring and scaling your Azure Batch pools.
   
5.  Now you are ready to start, please jump to [Run application](#run-application)

For details on what is configured see below.

## Set up NFS server and virtual network

A network file system (NFS) server is used as a shared storage for applications
that need to share data across message passing interface (MPI) processes. In the
GitHub repo, the deploy\_nfs.json file builds a virtual network and an NFS
server with a 1 TB data disk.

> **IMPORTANT**: This step is mandatory as virtual machine pools will be created
> within this virtual network.

The virtual network is created with these properties:

-   Default name: `hpcvnet`
-   Address space: `10.0.0.0/20`
-   Subnets:
    -   admin : `10.0.2.0/24`
    -   compute : `10.0.8.0/21`
-   NSG:
    -   admin : Allow SSH and HTTPS inbound connections

The NFS server is created with these properties:

-   Name : `nfsnode`
-   Private IP : `10.0.2.4`
-   Admin : `hpcadmin`
-   SSH Key : `~/.ssh/id_rsa.pub`
-   Data disk : `1 TB` (P30)
-   NFS mount : `/data`

**To set up the NFS server and network:**

1.  Create a resource group to host your virtual network and NFS server if
    needed:

```
    az group create --location <location> --name <resource_group>
```

For Azure Batch, the resource group by default is the one that includes your Batch account. 

2.  Deploy the virtual network and the NFS server.

```
    ./deploy_nfs.sh <config.json> <vnet_name> <resource_group>
```

The admin account created for the NFS server is named **hpcadmin** and your local public SSH key located here will be used for authentication. After the script is run, the full DNS name of the virtual machine is displayed.

## Create site-to-site VPN

If your HPC application needs a license server, make sure it can be accessed
from the virtual machines in the Batch pools. We used
site-to-site VPN to connect the license servers located in the different
subscriptions to the compute and batch resources. We considered using virtual
network peering, but found it impractical in this scenario since it needs
permission from the virtual network owners on both sides of the peering
connection.

We created subnets and gateways across two subscriptions that hosted three
resource groups and three virtual networks. We used a hub-spoke model to connect
the compute and batch resource groups and virtual networks to the ISV License
Server network. The connections were set up bidirectionally to allow for
troubleshooting.

To create a site-to-site VPN or set of site-to-site VPN tunnels between virtual
networks:

1.  Create a virtual network if you do not have one already with its
    corresponding address spaces and subnets. Repeat this again for the
    secondary virtual network and so on. Detailed instructions are located
    [here](https://docs.microsoft.com/azure/vpn-gateway/vpn-gateway-howto-vnet-vnet-resource-manager-portal).

2.  If you already have two or more virtual networks or have completed the
    creation phase above, do the following under each of the virtual networks
    you want to connect together using site-to site-VPNs.

    1.  Create a gateway subnet.

    2.  Specify the DNS server (optional).

    3.  Create a virtual network gateway.

3.  When all the gateways are completed, create your virtual network gateway
    connections [using the Azure
    portal](https://docs.microsoft.com/azure/vpn-gateway/vpn-gateway-howto-vnet-vnet-resource-manager-portal).

**NOTE:** These steps work only for virtual networks in the same subscription.
If your virtual networks are in different subscriptions, you must [use
PowerShell to make the
connection](https://docs.microsoft.com/azure/vpn-gateway/vpn-gateway-vnet-vnet-rm-ps).
However, if your virtual networks are in different resource groups in the *same*
subscription, you can connect them using the portal.

For more information about virtual network peering, see these resources:

-   [General availability: Global VNet
    Peering](https://azure.microsoft.com/updates/global-vnet-peering-general-availability/)
-   [Virtual network peering
    documentation](https://docs.microsoft.com/azure/virtual-network/virtual-network-peering-overview)
-   [Create, change, or delete a virtual
    network](https://docs.microsoft.com/azure/virtual-network/manage-virtual-network)

# Create Azure resources and configuration

All of the scripts used to build an image, create a cluster, and run
applications use the same configuration file, config.json. This file is an Azure
Resource Manager template in JavaScript Object Notation (JSON) format that
defines the resources to deploy. It also defines the dependencies between the
deployed resources. Before running the automation scripts, you must update the
config.json file with your values as explained in the sections below.

**config.json**

```json
{
    "subscription_id": "",
    "location": "",
    "packer": {
        "executable": "/bin/packer",
        "base_image": "baseimage.sh",
        "private_only": "false",
        "vnet_name": "",
        "subnet_name": "",
        "vnet_resource_group":""
    },
    "service_principal": {
        "tenant_id": "",
        "client_id": "",
        "client_secret": "<keyvaultname>.spn-secret"
    },
    "images": {
        "resource_group": "",
        "storage_account": "",
        "name": "centos74",
        "publisher": "OpenLogic",
        "offer": "CentOS-HPC",
        "sku": "7.4",
        "vm_size": "STANDARD_H16R",
        "nodeAgentSKUId": "batch.node.centos 7"
    },
    "cluster_type": "batch",
    "batch": {
        "resource_group": "",
        "account_name": "",
        "storage_account": "",
        "storage_container": "starttasks"
    },
    "infrastructure": {
        "nfsserver": "10.0.2.4",
        "nfsmount": "/data",
        "beegfs": {
            "mgmt":"x.x.x.x",
            "mount":"/beegfs"
        },
        "licserver": "",
        "network": {
            "resource_group": "",
            "vnet": "hpcvnet",
            "subnet": "compute",
            "admin_subnet": "admin"
        },
        "analytics": {
            "workspace":"",
            "key":"<keyvaultname>.analytics-key"
        }
    },
    "appstorage": {
        "resource_group": "",
        "storage_account": "",
        "app_container": "apps",
        "data_container": "data"
    },
    "results": {
        "resource_group": "",
        "storage_account": "",
        "container": "results"
    }
}
```

> **Note**: Secrets have to be stored in a KeyVault, see the section below on how to store your secrets
>

## Create a Key Vault to store secrets
Azure Key Vault is used to store your secret and avoid storing them in configuration files. Configuration fields storing secrets will instead store the keyvault name and the key name using the naming pattern **vault_name.key_name**. If you use the setup.sh script, the Key Vault will be automatically created for you and filled with the required secrets. 

To create a Key Vault use this command :

``` 
az keyvault create --name 'Contoso-Vault2' --resource-group 'ContosoResourceGroup' --location eastus
``` 

To add a secret to Key Vault use this command :

``` 
az keyvault secret set --vault-name 'Contoso-Vault2' --name 'ExamplePassword' --value 'Pa$$w0rd'
``` 

> **Note**: Make sure you are granted read access to the KeyVault Secrets. Check the Access policies of the Key Vault to see if your account is listed there with the read/list access to secrets.
>

To learn more about Key Vault see the online documentation [here](https://docs.microsoft.com/en-us/azure/key-vault/)


## Specify subscription and location

The region where deploy your resources must match the region used by your Batch
account.

Update the following lines in config.json:

```json
{
    "subscription_id": "your subscription id",
    "location": "region where to deploy your resources"
}
```

## Get Packer

You must copy the [Packer](https://www.packer.io/intro/index.html) binary and
add it to the `bin` directory of the repo. Then retrieve the full Packer path
and update the executable value in config.json as follows:

```json
"packer": {
    "executable": "bin/packer",
    "base_image": "baseimage-centos.sh",
    "private_only": "false",
    "vnet_name": "",
    "subnet_name": "",
    "vnet_resource_group":""
}
```

If you don't have one, create a service principal for Packer as follows:

```
az ad sp create-for-rbac --name my-packer
```

Outputs:

```json
{
    "appId": "99999999-9999-9999-9999-999999999999",
    "displayName": "my-packer",
    "name": "http://my-packer",
    "password": "99999999-9999-9999-9999-999999999999",
    "tenant": "99999999-9999-9999-9999-999999999999"
}
```

Using these output values, in config.json, update the values for **tenant_id**,
**client_id**. Store the **client_secret** in your KeyVault under the spn-secret key:

```json
"service_principal": {
    "tenant_id": "tenant",
    "client_id": "appId",
    "client_secret": "<keyvaultname>.spn-secret"
}
```

> By default Packer create a VM with a public IP and use it to SSH in and inject the configuration scripts. If you want to stay in a private network, you need to run the build_image.sh script within that VNET, set **private_only** to **true**, and specify values for **vnet_name**, **subnet_name** and **vnet_resource_group**.


## Store the images

To store the images, create a storage account in the same location as your Batch
account as follows:

```
az group create --location <location> --name my-images-here
az storage account create -n "myimageshere" -g my-images-here --kind StorageV2 -l <location> --sku Standard_LRS
```

Then specify the operating system, virtual machine size, and batch node agent to
be used. You can use the following command to list the agents and operating
systems supported by Azure Batch:

```
az batch pool node-agent-skus list --output table
```

To list the virtual machines sizes available in your region, run this command:

```
az vm list-sizes --location <region> --output table
```

In config.json, update the values in the **images** section:

```json
"images": {
    "resource_group": "my-images-here",
    "storage_account": "myimageshere",
    "name": "centos76",
    "publisher": "OpenLogic",
    "offer": "CentOS-HPC",
    "sku": "7.6",
    "vm_size": "STANDARD_HC44RS",
    "nodeAgentSKUId": "batch.node.centos 7"
}
```

## Specify the cluster type

```
    "cluster_type": "batch",
```

Specify the resource group containing the Batch account,
the name of the Batch account, and the name of the storage account linked to the
Batch account. If you don't have a Batch account, use these
[instructions](https://docs.microsoft.com/azure/batch/batch-account-create-portal)
to create one.

```json
"batch": {
    "resource_group": "batchrg",
    "account_name": "mybatch",
    "storage_account": "mybatchstorage",
    "storage_container": "starttasks"
}
```


## Identify infrastructure addresses

In config.json, you must specify the IP addresses of the NFS server and license
server. If you use the default scripts to build your NFS server, the IP address
is 10.0.2.4. The default mount is **/data**. For the **network** values, specify the **resource_group** name.
The values for **vnet**, **subnet**, and **admin_subnet** are the same as those
used earlier when building the NFS server and virtual network.

> **NOTE:** If you have an Avere system deployed in your network, you can specify its IP and mount point in the NFS settings.

If you have a BeeGFS system setup, you can specify the management IP and the mount point (default is **/beegfs**). By default the client version 7.0 is setup inside the images.

> **NOTE:** See this [repo](https://github.com/paulomarquesc/beegfs-template) for reference on how to deploy BeeGFS on Azure


```json
"infrastructure": {
    "nfsserver": "10.0.2.4",
    "nfsmount": "/data",
    "beegfs": {
        "mgmt":"x.x.x.x",
        "mount":"/beegfs"
    },
    "licserver": "w.x.y.z",
    "network": {
        "resource_group": "batchrg",
        "vnet": "hpcvnet",
        "subnet": "compute",
        "admin_subnet": "admin"
    }
}
```

## Set up analytics

If you want to store the telemetry data collected during the runs,
you can create an Azure [Log Analytics
workspace](https://docs.microsoft.com/azure/log-analytics/log-analytics-quick-create-workspace).
Once in place, update config.json and specify the Log Analytics workspace ID. Store the application key from your environment in KeyVault under **analytics-key** :

```json
"analytics": {
    "workspace":"99999999-9999-9999-9999-999999999999",
    "key":"<keyvaultname>.analytics-key"
}
```

The key can be find in the Advanced Settings / Connected sources of your workspace.

## Application storage

Application packages are stored in an Azure Storage account that uses one
container for the HPC application binaries and one container for the test
cases. Each HPC application is listed in a subfolder of the **apps** folder that
contains both binaries and data. Many applications are already included in the
repo, but if you do not see the one you want, you can add one as described
[later](#add-an-application-to-the-catalog) in this guide.

In config.json, update the following:

```json
"appstorage": {
    "resource_group": "myappstore resource group",
    "storage_account": "myappstore",
    "app_container": "apps",
    "data_container": "data"
}
```

## Store results

Application outputs and logs are automatically stored inside blobs within the
Storage account and container specified in config.json. Specify the location for
your results as follows:

```json
"results": {
    "resource_group": "myresults resource group",
    "storage_account": "myresults",
    "container": "results"
}
```

# Run application

To run a specific HPC application, do the following:

1.  Build an image for that application by running the following script,
    replacing **app-name** with the value documented in the table below:

```
    ./build_image.sh -a app-name
```

2.  Create a virtual machine pool for Batch running the following script, replacing **app-name** with the value documented in the [table](#set-up-an-hpc-application) below:

```
    ./create_cluster.sh -a app-name
```

3.  After the Azure Batch pool is created, scale it to the number of nodes you
    want to run on. To scale a
    pool, use Azure portal or [Batch Explorer](https://github.com/Azure/BatchExplorer).

4.  Run the HPC application you want as the following section
    describes.

```
    ./run_app.sh -a app-name \
            -c config -s script \
            -n nodes -p process_per_node \
            -x script_options
```


## Set up an HPC application 

Instructions for setting up, running, and analyzing results for the following
HPC applications are included in this repo, and more are being added. If the
application you want is not shown here, see the next section.

| **Application**                                                          | **app-name**       | **Versions**   | **Shared FS** | **Licence** |
|--------------------------------------------------------------------------|--------------------|----------------|---------------|-------------|
| [Abaqus](./documentation/apps/abaqus.md)                                 | abaqus             | 2017           |    No         |  Yes        |
| [ANSYS Mechanical](./documentation/apps/mechanical.md)                   | mechanical         | 19.2           |    No         |  Yes        |
| [ANSYS Fluent](./documentation/apps/fluent.md)                           | fluent             | 19.2           |    No         |  Yes        |
| [Builder](./documentation/apps/builder.md) - *build scripts*             | builder            |                |    No         |  No         |
| [Empty](./documentation/apps/empty.md) - *diagnostics*                   | empty              |                |    No         |  No         |
| [Gromacs](./documentation/apps/gromacs.md)                               | gromacs            | 2018.1         |    Yes        |  No         |
| [Linpack](./documentation/apps/empty.md)                                 | empty              | 2.3            |    No         |  No         |
| [NAMD](./documentation/apps/namd.md)                                     | namd               | 2.10           |    No         |  No         |
| [nwchem](./documentation/apps/nwchem.md)                                 | nwchem             | 6.8            |    Yes        |  No         |
| [OpenFOAM](./documentation/apps/openfoam.md)                             | openfoam           | 4.x            |    Yes        |  No         |
| [Pamcrash](./documentation/apps/pamcrash.md)                             | pamcrash           | 2017 2018      |    No         |  Yes        |
| [Quantum Espresso](./documentation/apps/qe.md)                           | qe                 | 6.3            |    Yes        |  No         |

## Add an application to the catalog

Many common HPC applications are included in the repo’s catalog in the **apps**
folder. To add an application to the catalog:

1.  Under **apps**, create a new folder with the name of the application. Use
    only the lowercase letters *a* to *z*. Do not include numbers, special
    characters, or capital letters.

2.  Add a file to the folder called `install_<app-name>.sh`. The folder name
    must match the name used by `create_image.sh` to call
    `install_<app-name>.sh`.

3.  In the new `install_<app-name>.sh` file, add steps to copy the executables

changelog.md
### Version 1.8
- Fix setup.sh
- Added Platform MPI support for Fluent
- Support CentOS-HPC 7.6 image
- Removed CycleCloud
- Added modules for Intel MPI 2018
- Use modules in the empty app
- Added Pamcrash

### Version 1.7
- Added Centos76 support for HC/HB
- Removed Centos75 support for HC/HB
- Force packer 1.3.2 as 1.3.4 has a bug not using Premium_LRS for captured image
- Added QE 6.3
- Resolved: Use HTTPS instead of HTTP when accessing DATA storage
- Resolved: Improve ERROR parsing from the output of image craeation 
- Allow packer to use private IP only
- Force waagent to version 2.2.36 which fix the IB0 not being set for HC/HB


### Version 1.6
- Support HC44rs VM Type
- Added builder application to build OSS applications
- Added -u option for run_app to specify a Batch autoUser/scope/elevation value or a specific username
- disable GSS proxy in the base image script for CentOS
- Moved build_openmpi from empty to builder
- Added build_hplamd, build_stream (for HB/HC) script to builder
- Added the VMSIZE, HPC_APPS_STORAGE_ENDPOINT, HPC_APPS_SASKEY, UCX_IB_PKEY, IB_PKEY environment variables available when executing run scripts
- Surround SCRIPT_FLAGS with " in task-template.json to allow passing SAS_KEY in params of scripts
- Updated empty/run_single_hpl_epyc to use hpl built from build_hplamd and stored into blobs
- Added run_stream script to empty that support HB and HC
- Update WAAGENT when building the image
- Upgrade LIS to version 4.2.8
- disable cpupower and firewalld for all SKUs
- use version 1.3.2 for packer as 1.3.3 has a bug for managed image


### Version 1.5
- Use Key Vault to store secrets instead of clear values in configuration files
- Support HB60rs VM Type
  - Setup Mellanox drivers
  - Setup IB
  - Setup Intel MPI 2018
- Use managed disks for OS disks with Packer and Batch for VM type that support it
- Upgraded to Fluent 19.2
- Added new test cases in the empty image (allreduce, ring_pingpong, all_pingpong, stream, single_hpl_epyc)
- new -u option for create_cluster to update a pool if it exists
- new -r option for run_app to repeat task n times
- New -q option for run_app to use a specific queue (pool), this allow running diagnostics tests on existing application pools

### Version 1.4
- Added nwchem
- Update documentation to use setup.sh to configure config file
- Update documentation for app binaries and data upload location

### Version 1.3
- Added OpenFOAM 4.x
- Added Abaqus 2017
- Added NAMD 2.10
- Fix broken links in documentation

### Version 1.0
Initial release

setup.sh
#!/bin/bash
DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
source "$DIR/common.sh"

TIMESTAMP=$(date +"%Y%m%d%H%M")
LOGDIR=log-$TIMESTAMP
mkdir $LOGDIR

config_file="config.json"
debug_mode=0

function usage {
    me=`basename $0`
    echo ""
    echo "Usage:"
    echo "$me -c config.json -d"
    echo "    -c: config file to use"
    echo "    -d: debug mode (leave log files)"
    echo "    -h: help"
    echo ""
}

while getopts "c: d h" OPTION
do
    case ${OPTION} in
        c)
        config_file=$OPTARG
            ;;
        d)
        debug_mode=1
            ;;
        h)
        usage
        exit 0
            ;;
    esac
done

if [ ! -e "$config_file" ]; then
    cp template-config.json $config_file
fi

function is_not_set {
    val="$(jq -r $1 $config_file)"
    [ "$val" = "NOT-SET" -o "$val" = "" ]
}

function create_storage_account {
    local sa=$1
    local rg=$2
    local loc=$3

    exists=$(az storage account list -g $rg -o table | grep $sa)

    if [ "$exists" = "" ]; then
        echo "Creating storage account ($sa)"
        az storage account create \
            --name $sa \
            --sku Standard_LRS \
            --kind StorageV2 \
            --location $loc \
            --resource-group $rg 
        if [ $? != 0 ]; then
            echo "ERROR: Failed to storage account"
            exit 1
        fi
    fi
}

jqstr="."

function update_config_file {
    tag=$1

    if [ "$jqstr" != "." ]; then
        if [ "$debug_mode" -eq "1" ]; then
            echo "Applying changes: $jqstr"
        fi

        mv $config_file $LOGDIR/${config_file%.*}-${TIMESTAMP}-${tag}.json
        jq "$jqstr" $LOGDIR/${config_file%.*}-${TIMESTAMP}-${tag}.json > ${config_file}

        jqstr="."
    else
        if [ "$debug_mode" -eq "1" ]; then
            echo "No changes to apply."
        fi
    fi
}

read_value subscription_id .subscription_id
if is_not_set .subscription_id; then
    az account list > $LOGDIR/az_account_list.log
    jq -r  '.[] | "\(.id) (\(.name))"' $LOGDIR/az_account_list.log
    echo -n "Enter the subscription id to use: "
    read subscription_id
    jqstr+="|.subscription_id=\"$subscription_id\""

    update_config_file subscription
fi

echo "Setting to correct subscription id"
az account set --subscription $subscription_id > $LOGDIR/az_account_set.log

read_value location .location
if is_not_set .location; then
    echo -n "Enter location (for anything new to be created in): "
    read location
    jqstr+="|.location=\"$location\""

    update_config_file location
fi

read_value resource_group .resource_group
if is_not_set .resource_group; then
    echo -n "Enter resource group (for anything new to be created in): "
    read resource_group

    group_exists=$(az group exists -n $resource_group)
    if [ "$group_exists" = "false" ];then
        az group create --name $resource_group --location "$location"
    fi

    jqstr+="|.resource_group=\"$resource_group\""

    update_config_file resource_group
fi
echo ""
echo "** Note: Azure Key Vault is used to store secrets"
echo "   If you specify an existing Vault, this script will add secrets in it"
echo "   If the Vault doesn't exists, it will create one for you"

echo -n "Enter the Key Vault Name: "
read vault_name
vault=$(az keyvault show --name $vault_name --output tsv)
if [ "$vault" = "" ]; then
    az keyvault create --name $vault_name --resource-group $resource_group
fi

if is_not_set .service_principal.name; then
    echo -n "Enter the service principal name to create: "
    read sp_name
    az ad sp create-for-rbac --name $sp_name > $LOGDIR/az_ad_sp_create-for-rbac.log

    secret=$(jq -r '.password' $LOGDIR/az_ad_sp_create-for-rbac.log)
    az keyvault secret set --vault-name $vault_name --name "spn-secret" --value $secret
    
    jqstr+="|.service_principal.name=\"$sp_name\""
    jqstr+="|.service_principal.tenant_id=\"$(jq -r '.tenant' $LOGDIR/az_ad_sp_create-for-rbac.log)\""
    jqstr+="|.service_principal.client_id=\"$(jq -r '.appId' $LOGDIR/az_ad_sp_create-for-rbac.log)\""
    jqstr+="|.service_principal.client_secret=\"$vault_name.spn-secret\""

    update_config_file service_principal
fi

if is_not_set .packer.executable; then
    packer_version=1.3.2
    if [[ $(uname -r) =~ Microsoft$ ]]; then
        echo "Downloading Windows version of Packer"
        packerzip=https://releases.hashicorp.com/packer/$packer_version/packer_${packer_version}_windows_amd64.zip
        jqstr+="|.packer.executable=\"bin/packer.exe\""
    else
        echo "Download Linux version of Packer"
        packerzip=https://releases.hashicorp.com/packer/$packer_version/packer_${packer_version}_linux_amd64.zip
        jqstr+="|.packer.executable=\"bin/packer\""
    fi
    
    mkdir -p bin && cd bin \
        && wget $packerzip -O packer.zip \
        && unzip packer.zip && rm -f packer.zip \
        && cd -

    update_config_file packer
fi


if is_not_set .images.name; then
    echo "Enter image/SKU to use:"
    echo " - centos74    for H16r"
    echo " - centos76_hb for HB60rs"
    echo " - centos76_hc for HC44rs"
    read image
    valid_image=1
    case "$image" in
        centos76_hb)
            packer_base_image="baseimage-centos.sh"
            images_name="centos76"
            images_publisher="OpenLogic"
            images_offer="CentOS"
            images_sku="7.6"
            images_vm_size="Standard_HB60rs"
            images_nodeAgentSKUId="batch.node.centos 7"
            ;;
        centos76_hc)
            packer_base_image="baseimage-centos.sh"
            images_name="centos76"
            images_publisher="OpenLogic"
            images_offer="CentOS"
            images_sku="7.6"
            images_vm_size="Standard_HC44rs"
            images_nodeAgentSKUId="batch.node.centos 7"
            ;;
        centos74)
            packer_base_image="baseimage-centos.sh"
            images_name="centos74"
            images_publisher="OpenLogic"
            images_offer="CentOS-HPC"
            images_sku="7.4"
            images_vm_size="STANDARD_H16R"
            images_nodeAgentSKUId="batch.node.centos 7"
            ;;
        ubuntu1604)
            packer_base_image="baseimage-ubuntu.sh"
            images_name="ubuntu1604"
            images_publisher="Canonical"
            images_offer="UbuntuServer"
            images_sku="16.04-LTS"
            images_vm_size="STANDARD_NC24r"
            images_nodeAgentSKUId="batch.node.ubuntu 16.04"
            ;;
        *)
            echo "Unrecognised option ($image), skipping."
            valid_image=0
    esac
    if [ "$valid_image" = "1" ]; then
        jqstr+="|.packer.base_image=\"$packer_base_image\""
        jqstr+="|.images.name=\"$images_name\""
        jqstr+="|.images.publisher=\"$images_publisher\""
        jqstr+="|.images.offer=\"$images_offer\""
        jqstr+="|.images.sku=\"$images_sku\""
        jqstr+="|.images.vm_size=\"$images_vm_size\""
        jqstr+="|.images.nodeAgentSKUId=\"$images_nodeAgentSKUId\""
    fi

    update_config_file image_and_sku
fi

if is_not_set .images.storage_account; then
    echo -n "Enter the storage account to be created for images: "
    read sa
    create_storage_account $sa $resource_group $location
    jqstr+="|.images.storage_account=\"$sa\""
    jqstr+="|.images.resource_group=\"$resource_group\""

    update_config_file images_storage
fi

read_value cluster_type .cluster_type
if is_not_set .cluster_type; then
    cluster_type="batch"

    jqstr+="|.cluster_type=\"$cluster_type\""

    update_config_file cluster_type
fi

if [ "$cluster_type" = "batch" ]; then

    # check batch account and set if necessary
    read_value batch_storage_account .batch.storage_account
    if is_not_set .batch.storage_account; then
        echo -n "Enter batch storage account name: "
        read batch_storage_account
        create_storage_account $batch_storage_account $resource_group $location
        jqstr+="|.batch.storage_account=\"$batch_storage_account\""

        update_config_file batch_storage
    fi

    if is_not_set .batch.account_name; then
        echo -n "Enter batch account name: "
        read batch_account_name
        az batch account create \
            --location $location \
            --name $batch_account_name \
            --resource-group $resource_group \
            --storage-account $batch_storage_account \
            > $LOGDIR/az_batch_account_create.log
        if [ $? != 0 ]; then
            echo "ERROR: Failed to create batch account"
            exit 1
        fi

        jqstr+="|.batch.resource_group=\"$resource_group\""
        jqstr+="|.batch.account_name=\"$batch_account_name\""

        update_config_file batch_account
    fi
    
else
    echo "Unknown cluster type ($cluster_type)."
fi

# look at infrastructure
if is_not_set .infrastructure.network.vnet; then
    echo "Infrastrure (network, NFS) node defined.  Setting up infrastructure".
    rsaPublicKey=$(cat $HOME/.ssh/id_rsa.pub)
    if [ "$rsaPublicKey" == "" ]; then    
        echo "No public ssh key found (in ~/.ssh/id_rsa.pub)."
        echo "Run ssh-key with all default options and re-run this script."
        exit 1
    fi

    echo -n "Enter VNET name to set up infrastructure in: "
    read vnetname
    jqstr+="|.infrastructure.network.vnet=\"$vnetname\"" 

    parameters=$(cat << EOF
{
    "vnetName": {
        "value": "$vnetname"
    },
    "rsaPublicKey": {
        "value": "$rsaPublicKey"
    }
}
EOF
)

    echo "Deploying VNET $vnetname and NFS Server in resource group $resource_group "
    az group deployment create \
        --resource-group "$resource_group" \
        --template-file infrastructure/deploy_nfs.json \
        --parameters "$parameters" | tee $LOGDIR/deploy_nfs.log

    nfsnode=$(jq -r '.properties.outputs.nfsname.value' $LOGDIR/deploy_nfs.log)
    echo "NFS server can be reach at $nfsnode"
    echo "uploading setup script"
    # remove old entries for the nfs nodes in case of several retries to avoid ssh failures
    ssh-keygen -f "~/.ssh/known_hosts" -R ${nfsnode}
    scp -o StrictHostKeyChecking=no infrastructure/setup_nfs.sh hpcadmin@${nfsnode}:setup_nfs.sh
    echo "Configuring NFS disk and mount point"
    ssh -o StrictHostKeyChecking=no hpcadmin@${nfsnode} "sudo bash setup_nfs.sh"

    jqstr+="|.infrastructure.nfsserver=\"10.0.2.4\""
    jqstr+="|.infrastructure.licserver=\"10.1.0.4\""
    jqstr+="|.infrastructure.network.resource_group=\"$resource_group\""
    jqstr+="|.infrastructure.network.vnet=\"$vnetname\""
    jqstr+="|.infrastructure.network.subnet=\"compute\""
    jqstr+="|.infrastructure.network.admin_subnet=\"admin\""

    update_config_file infrastructure
fi

# analytics
if is_not_set .infrastructure.analytics.workspace; then
    echo "Warning: Log analytics not set up."
    echo "         To use this feature, set up an account and set workspace/key."
fi

# app storage
if is_not_set .appstorage.storage_account; then
    echo "Warning: You do not have anything set for app storage."
    echo "         You must either create a copy installers/test data."
    echo "         Alternatively obtain account details and SAS key existing one."

    echo -n "Enter the storage account to be created for apps: "
    read sa
    create_storage_account $sa $resource_group $location
    jqstr+="|.appstorage.storage_account=\"$sa\""
    jqstr+="|.appstorage.resource_group=\"$resource_group\""
    
    update_config_file apps_storage
fi

if is_not_set .results.storage_account; then
    echo -n "Enter the storage account to be created for results: "
    read sa
    create_storage_account $sa $resource_group $location
    jqstr+="|.results.storage_account=\"$sa\""
    jqstr+="|.results.resource_group=\"$resource_group\""
    
    update_config_file results_storage
fi

if [ "$debug_mode" -eq "0" ]; then
    rm -rf $LOGDIR
fi
